README.md
configured-swagger-client
=========================

[![wercker status](https://app.wercker.com/status/15fcfccb064836e35c8438135e37cd41/s/master "wercker status")](https://app.wercker.com/project/byKey/15fcfccb064836e35c8438135e37cd41)


GasBuddy has chosen [Swagger](http://swagger.io/) as our API format for all services. Luckily, a growing number of
partners are also using Swagger. Our services are written mostly in Node.js, and this module is an attempt to make the
process of wiring a Node.js project to Swagger services easier and more robust.

We have a variety of "independent components":

* -api - a Swagger API that is reachable by the outside world
* -serv - a Swagger API that is only reachable from other services
* -web - a non-Swagger service that exposts
* -job - a start-and-finish job that runs periodically or on demand

All of these components likely make use of other Swagger services. Each of these services is defined by a JSON swagger
specification and will use [small-swagger-codegen](https://github.com/gas-buddy/small-swagger-codegen) to generate a
client package. This is a departure from versions < 5.x of this library which generated a JSON spec and the
client was dynamically generated. THe value of pregeneration is full type information available at development time,
which includes autocompleting complex method names and request structures, which in a language like Javascript, can reduce
bugs significantly without losing the flexibility (since it's just a generator doing the typing work).

One of the other important things we do is create a "CorrelationId" - a single identifier which can be traced
all through the logs of all of the services that request touches. This module creates an easy way to pass that
CorrelationId along in dependent service calls.

See the tests for example usage, but here's a simple one:

```
import PetApi from 'a-node-module-from-small-swagger-codegen';

const petApi = new PetApi(req.gb.serviceFactory);

const pets = await services.Pets.findPetsByStatus({ status: 'pending' });
// pets.obj has your pets!
```

API Options
===========

The CorrelationId support is accomplished via a requestInterceptor option to the rest-api-support methods.
rest-api-support also supports adding headers (such as authorization headers) in this phase. Metric collection is also
enabled by the requestInterceptor/responseInterceptor pair.


CHANGELOG.md
1.2.2
=====
* Don't throw an exception if you ask for a non-existent method name (just return null)

1.2.1
=====
* Support promises properly, add test to illustrate

1.2.0
=====
* Add a preProcessor option for modifying swagger configuration
* Support promises for both preProcessor and postProcessor

1.1.0
=====
* Change postProcessor function signature to get all properties as an object. Strictly, this
is a violation of semver. Realistically, it's been like an hour, so I will unpublish the previous.

1.0.1
=====
* Add CHANGELOG.md
* Fix README.md markdown for links

1.3.0
=====
* Add simple class to allow direct hydration from a (to-be-released) configuration layer
* Added 'enabled' property on endpoint config that, when false, will not create the specified service client

1.4.0
=====
* Support chaining requestInterceptor and responseInterceptor so we don't drop handlers on the floor.

2.0.0
=====
* Supports node 8 only

2.2.0
=====
* Add a symbol and a property using that symbol to the objects passed as "this" to request/response interceptors. The value of this property is an Error object whose stack was captured at the BEGINNING of the Swagger request, which often makes a more meaningful trace.

2.3.1
=====
* Fix error symbol when options are not passed (meaning, like all the time)

5.0.0
======
* A complete rewrite using pre-generated swagger clients which significantly simplifies essentially all parts of this module

Migration:
----------
* There are no more tag names in swagger specs currently, so that's a big change
* Instead of `req.gb.services.FooServ.default.method_name()` it is now:
```
const fooServ = new FooServ(req);
fooServ.method_name();
```
* In your config.json(s), change "services" to "serviceFactory", and change "specs" to "clients", and then update all the package names to point to the generated-code version (e.g. feature-api-spec becomes feature-api-js-client)
* No more "expects" on the promise, just "expect" - so canonicalize that

test/test_clients.js
import tap from 'tap';
import nock from 'nock';
import { EventEmitter } from 'events';
import Client from '../src/index';
import FeatureApi from './swagger';

const service = new EventEmitter();

const client = new Client({ service }, {
  clients: {
    'feature-api': FeatureApi,
  },
  endpoints: {
    'feature-api': {
      protocol: 'http',
      port: 1234,
    },
  },
});

class DerivativeApi extends FeatureApi {
  constructor(config) {
    super(config(DerivativeApi));
  }
}

service.serviceFactory = client.start({ service });
service.name = 'test-serv';
const featureApi = new FeatureApi(service.serviceFactory);

const fakeRequest = { app: { id: 'GasBuddy' } };

tap.test('test_clients', async (tester) => {
  tester.test('simple call', async (t) => {
    nock('http://feature-api:1234')
      .post('/feature/features/foobar', { app: { id: 'GasBuddy' } })
      .reply(200, {});

    const response = await featureApi.getFeatures({
      tag: 'foobar',
      client: fakeRequest,
    });
    t.strictEquals(response.status, 200, 'Should find pets');
    t.ok(response.body, 'Should return a body');
  });

  tester.test('interception', async (t) => {
    t.plan(2);
    nock('http://feature-api:1234')
      .post('/feature/features/foobar', { app: { id: 'GasBuddy' } })
      .reply(200, {});
    await featureApi.getFeatures({
      tag: 'foobar',
      client: fakeRequest,
    }, {
      requestInterceptor() {
        t.ok(true, 'Should call request interceptor');
      },
      responseInterceptor() {
        t.ok(true, 'Should call response interceptor');
      },
    });
  });

  tester.test('exception handling should work', async (t) => {
    nock('http://feature-api:1234')
      .post('/feature/features/baz', { app: { id: 'GasBuddy' } })
      .reply(404, {});

    t.plan(3);
    client.once('error', () => {
      t.ok('Should call error event handler');
    });
    const { status } = await featureApi.getFeatures({
      tag: 'baz',
      client: fakeRequest,
    }).expect(404);
    t.ok(true, 'Should not throw for expected error');
    t.strictEquals(status, 404, 'Should return 404');
  });

  tester.test('Incorrect service factory should throw', (t) => {
    t.throws(() => new DerivativeApi(service.serviceFactory), 'Unknown service should throw');
    t.end();
  });

  tester.test('HTTPS should work with logging', async (t) => {
    const secureClientLogger = new Client({ service }, {
      clients: {
        'feature-api': FeatureApi,
      },
      endpoints: {
        'feature-api': {
          protocol: 'https',
          log: true,
        },
      },
    });
    service.serviceFactory = secureClientLogger.start({ service });
    const secureApi = new FeatureApi(service.serviceFactory);

    nock('https://feature-api:8443')
      .post('/feature/features/foobar', { app: { id: 'GasBuddy' } })
      .reply(200, {});

    const response = await secureApi.getFeatures({
      tag: 'foobar',
      client: fakeRequest,
    });

    t.strictEquals(response.status, 200, 'Should get a 200');
  });
});


test/swagger/index.js
// eslint-disable-next-line no-unused-vars
import { parameterBuilder, fetchHelper, eventSourceHelper } from 'rest-api-support';

/**
 *
 * @export
 * @class FeatureAPI
 */
export default class FeatureAPI {
  constructor(configOrGenerator) {
    let config = configOrGenerator;
    if (typeof configOrGenerator === 'function') {
      config = configOrGenerator(FeatureAPI);
    }
    const {
      baseUrl = '',
      fetch,
      EventSource,
      requestInterceptor,
      responseInterceptor,
    } = config || {};
    Object.assign(this, { baseUrl, fetch, requestInterceptor, responseInterceptor, EventSource });
  }

  /**
   * Get a list of features and settings for a given device, user and app
   *
   * @parameter { string } tag: The tag (and its parents) for which features are being requested
   * @parameter { string } sampleQuery: A query parameter
   * @parameter { ClientData } client: Information about the client making the request
   */
  getFeatures({
    tag,
    sampleQuery,
    client,
  }, fetchOptions) {
    // Build parameters, run request interceptors, fetch, and then run response interceptors
    // eslint-disable-next-line prefer-rest-params
    const source = { method: 'getFeatures', client: '', args: arguments[0] };
    const fetchArgs = parameterBuilder('POST', this.baseUrl, '/feature/features/{tag}')
      .path('tag', tag)
      .query('sample_query', sampleQuery)
      .body('client', client)
      .build();
    return fetchHelper(this, fetchArgs, fetchOptions, source);
  }
}


test/test_configuration.js
import tap from 'tap';
import path from 'path';
import * as clientConfig from '../src/index';
import part1 from './swagger/part1.json';
import sample from './swagger/sample.json';

tap.test('Should work with no services', async (t) => {
  const ret = await clientConfig.configureServices({});
  t.deepEquals(ret, {}, 'Should not return any services');
  t.end();
});

tap.test('Should work with a JSON config', async (t) => {
  const ret = await clientConfig.configureServices({
    test: part1,
  });
  t.ok(ret.Test, 'Should configure the Test service');
  t.ok(ret.Test.apis.default, 'Test service should have a default tag');
  t.ok(ret.Test.apis.default.get_goodbye, 'Test service should have known API');
  t.strictEquals(ret.Test.url, 'https://test:8443', 'Default host should match');
  t.end();
});

tap.test('Should work with an endpoint', async (t) => {
  const ret = await clientConfig.configureServices({
    test: part1,
  }, {
    test: {
      hostname: 'foobar',
      protocol: 'http:',
      port: 1234,
    },
  });
  t.ok(ret.Test, 'Should configure the Test service');
  t.ok(ret.Test.apis.default, 'Test service should have a default tag');
  t.ok(ret.Test.apis.default.get_goodbye, 'Test service should have known API');
  t.strictEquals(ret.Test.url, 'http://foobar:1234', 'Default host should match');
  t.end();
});

tap.test('Should work with a URL', async (t) => {
  const ret = await clientConfig.configureServices({
    pets: 'http://petstore.swagger.io/v2/swagger.json',
  });
  t.ok(ret.Pets, 'Should configure the Pets service');
  t.ok(ret.Pets.apis.pet, 'Pet service should have a pet tag');
  t.ok(ret.Pets.apis.pet.findPetsByStatus, 'Pets service should have known API');
  t.strictEquals(ret.Pets.url, 'http://petstore.swagger.io/v2/swagger.json', 'Default host should match');
  t.end();
});

tap.test('Should work with a composite document', async (t) => {
  const ret = await clientConfig.configureServices({
    test: sample,
  }, {}, { basedir: path.join(__dirname, 'swagger') });
  t.ok(ret.Test, 'Should configure the Test service');
  t.ok(ret.Test.apis.default, 'Test service should have a default tag');
  t.ok(ret.Test.apis.default.get_hello_world, 'Test service should have known API');
  t.strictEquals(ret.Test.url, 'https://test:8443', 'Default host should match');
  t.end();
});

tap.test('Should pre and post process', async (t) => {
  t.plan(7);
  const ret = await clientConfig.configureServices({
    'test-serv': part1,
  }, {}, {
    preProcessor(info) {
      t.strictEquals(info.name, 'test-serv', 'Name should match');
      t.strictEquals(info.memberName, 'TestServ', 'Member name should match');
      t.strictEquals(info.config.url, 'https://test-serv:8443');
    },
    postProcessor(info) {
      t.strictEquals(info.name, 'test-serv', 'Name should match');
      t.strictEquals(info.memberName, 'TestServ', 'Member name should match');
      t.ok(info.client, 'Should have a client');
    },
  });
  t.ok(ret.TestServ, 'Client should be returned');
  t.end();
});

tap.test('Should work from the root module', async (t) => {
  const Constructor = clientConfig.default;
  const c = new Constructor({}, {
    specs: {
      test: part1,
    },
  });
  const final = await c.start();
  t.ok(final.Test, 'Should create services');
});


src/index.js
import defaultFetch from 'node-fetch';
import EventSource from 'eventsource';
import { EventEmitter } from 'events';

const CONFIG_FUNCTION = Symbol.for('small-swagger-codegen::configurationGenerator');
const CALLINFO = Symbol('Swagger call info key');
const logEverything = !!process.env.LOG_SWAGGER_CALLS;

function findServiceName(clientClass, clients) {
  const matchedEntry = Object.entries(clients)
    .find(([, client]) => (client?.default === clientClass || client === clientClass))
    || Object.entries(clients)
      .find(([, client]) => (client?.default?.name === clientClass.name || client.name === clientClass.name));
  return matchedEntry?.[0];
}

function serviceFactory(swaggerConfigurator, req) {
  const { config } = swaggerConfigurator;
  return (clientClass) => {
    // Our job is to build the configuration for the client.
    // First, find out which client this is.
    const serviceName = findServiceName(clientClass, config.clients);
    // If we don't know, throw an error
    if (!serviceName) {
      const err = new Error('Unknown service called');
      err.client = clientClass;
      throw err;
    }
    const { basePath = '', hostname = serviceName, port, protocol = 'https', noTracing, log, username, password, authToken } = config.endpoints[serviceName] || {};
    let newSpanLogger;
    const clientConfig = {
      fetch: config.fetch,
      EventSource,
      requestInterceptor(request, source) {
        source[CALLINFO] = {
          client: clientClass,
          serviceName,
          operationName: source.method,
          request,
        };
        request.headers = request.headers || {};
        if (!noTracing) {
          request.headers.correlationid = req.headers?.correlationid;
          newSpanLogger = req.gb?.logger?.loggerWithNewSpan?.();
          request.headers.span = newSpanLogger?.spanId;
        }
        if ((log || logEverything) && req.gb?.logger) {
          req.gb.logger.info('api-req', {
            url: request.url,
            m: request.method,
            childSp: newSpanLogger?.spanId,
          });
        }
        if (username) {
          const based = Buffer.from(`${username}:${password}`).toString('base64');
          request.headers.authorization = `Basic ${based}`;
        } else if (authToken) {
          request.headers.authorization = `Bearer ${authToken}`;
        }
        swaggerConfigurator.emit('start', source[CALLINFO]);
      },
      responseInterceptor(response, request, source) {
        const { status } = response;
        source[CALLINFO].response = response;
        if ((log || logEverything) && req.gb?.logger) {
          req.gb.logger.info('api-res', {
            s: response.status,
            l: response.headers['content-length'] || 0,
            childSp: newSpanLogger?.spanId,
          });
        }
        if (status >= 200 && status <= 300) {
          swaggerConfigurator.emit('finish', source[CALLINFO]);
        } else {
          swaggerConfigurator.emit('error', source[CALLINFO]);
        }
      },
    };
    const isHttps = protocol.toLowerCase().startsWith('https');
    const finalPort = Number(port || (isHttps ? 8443 : 8000));
    const portString = ((isHttps && finalPort === 443) || (!isHttps && finalPort === 80)) ? '' : `:${finalPort}`;
    clientConfig.baseUrl = `${protocol}${protocol.endsWith(':') ? '//' : '://'}${hostname}${portString}${basePath}`;
    return clientConfig;
  };
}

export default class SwaggerClientConfigurator extends EventEmitter {
  constructor(context, config) {
    super();
    this.config = {
      fetch: defaultFetch,
      ...config,
    };
    context.service.on('request', (req) => {
      const factory = serviceFactory(this, req);
      req[CONFIG_FUNCTION] = factory;
      req.gb.serviceFactory = factory;
    });
  }

  start(context) {
    const globalFactory = serviceFactory(this, {
      headers: {
        correlationid: `${context.service.name}-global`,
      },
    });
    // Provide to the event emitter methods to the containing service
    globalFactory.events = this;
    context.service[CONFIG_FUNCTION] = globalFactory;
    return globalFactory;
  }
}


src/proxy.js
export const OriginalCallPropertyKey = Symbol('An error object that is created at original call time for a Swagger call');
export const CallPathPropertyKey = Symbol('The configured service name and method used in the current call');

function chainInterceptors(defaultOptions, explicitOptions, placeholderError, opInfo) {
  const combined = Object.assign({}, defaultOptions, explicitOptions);
  combined.requestInterceptor = function combinedRequestInterceptor(...args) {
    this[OriginalCallPropertyKey] = placeholderError;
    this[CallPathPropertyKey] = opInfo;
    if (explicitOptions && explicitOptions.requestInterceptor) {
      explicitOptions.requestInterceptor.apply(this, args);
    }
    if (defaultOptions && defaultOptions.requestInterceptor) {
      defaultOptions.requestInterceptor.apply(this, args);
    }
  };
  if (defaultOptions && explicitOptions
    && defaultOptions.responseInterceptor
    && explicitOptions.responseInterceptor) {
    combined.responseInterceptor = function combinedResponseInterceptor(...args) {
      explicitOptions.responseInterceptor.apply(this, args);
      defaultOptions.responseInterceptor.apply(this, args);
    };
  }
  return combined;
}

export function servicesWithOptions(serviceCollection, options) {
  // This proxy function is used for each API on each service
  const apiHandler = {
    get(target, key) {
      if (target[key]) {
        const returnFunction = (params, explicitOptions) => {
          const placeholderError = new Error();
          Error.captureStackTrace(placeholderError, returnFunction);
          let defaultOptions = options;
          if (typeof options === 'function') {
            defaultOptions = options(key, params, explicitOptions);
          }
          const opInfo = [target[CallPathPropertyKey], key];
          const finalOptions = chainInterceptors(
            defaultOptions,
            explicitOptions,
            placeholderError,
            opInfo,
          );

          // Provide a convenience method on the promise
          return Object.assign(
            target[key](params, finalOptions), {
              expect(...codes) {
                return this.catch((error) => {
                  if (codes.includes(error.status)) {
                    return {
                      errObj: error,
                      response: error.response,
                      status: error.status,
                      statusCode: error.statusCode,
                      body: error.response && error.response.body,
                    };
                  }
                  throw error;
                });
              },
              expects(...codes) {
                return this.catch((error) => {
                  if (codes.includes(error.status)) {
                    return {
                      errObj: error,
                      response: error.response,
                      status: error.status,
                      statusCode: error.statusCode,
                      body: error.response && error.response.body,
                    };
                  }
                  throw error;
                });
              },
            },
          );
        };
        return returnFunction;
      }
      return target[key];
    },
  };

  // This proxy function is used for each service
  const serviceHandler = {
    get(target, key) {
      if (typeof target[key] === 'object') {
        const newProxy = new Proxy(target[key], apiHandler);
        newProxy[CallPathPropertyKey] = target[CallPathPropertyKey];
        return newProxy;
      }
      return target[key];
    },
  };

  const clientHandler = {
    get(target, key) {
      if (key === 'apis') {
        const newProxy = new Proxy(target.apis, serviceHandler);
        newProxy[CallPathPropertyKey] = target[CallPathPropertyKey];
        return newProxy;
      }
      if (typeof target.apis[key] === 'object') {
        const newProxy = new Proxy(target.apis[key], apiHandler);
        newProxy[CallPathPropertyKey] = target[CallPathPropertyKey];
        return newProxy;
      }
      return target[key];
    },
  };

  const proxied = {};
  // And this little proxy intercepts serviceCollection
  const collectionProxy = {
    get(target, key) {
      const exist = proxied[key];
      if (exist) {
        return exist;
      }
      const service = target[key];
      if (service) {
        // I don't think this actually has to get done every
        // time because we're setting the prop on the REAL
        // object, but it's so cheap that I'm leaving it simple.
        const newProxy = new Proxy(service, clientHandler);
        newProxy[CallPathPropertyKey] = key;
        proxied[key] = newProxy;
        return newProxy;
      }
      return service;
    },
  };

  return new Proxy(serviceCollection, collectionProxy);
}


src/configure.js
import _ from 'lodash';
import URL from 'url';
import path from 'path';
import Client from 'swagger-client';
import jsonResolver from '@gasbuddy/swagger-ref-resolver';

/**
 * Find the best protocol scheme (http/https) for a given GasBuddy service
 */
function scheme(endpoints, swagger) {
  if (endpoints.default && endpoints.default.protocol) {
    return endpoints.default.protocol;
  }
  if (Array.isArray(swagger.schemes)) {
    if (swagger.schemes.includes('https')) {
      return 'https:';
    }
    return 'http:';
  }
  return 'https:';
}

/**
 * Take a set of services and endpoint configurations for those
 * services and return an object with the same keys as services
 * and configured swagger clients as values
 * @param {object} services The services to be created. The key should be the name of the service
 *  (the first letter of this will always be made upper and camel cases in the returned map)
 * @param {object} endpoints Any endpoint configuration for the services. The key should match
 *  the services parameter key, and the values include hostname, port, protocol, basePath,
 *  baseReferencePath. If the value is a string, it is assumed to be the URL of the target service.
 *  In absence of an endpoint configuration, the name of the service will be used as the name of the
 *  host.
 * @param {object} [options] Any options for all services
 * @param {string} options.basedir This module uses swagger-ref-resolver to resolve "enhanced"
 *  swagger documents. Any relative paths in those documents will use basedir
 * @param {function} options.postProcessor A function that will be called with information
 *  about the client (name, memberName, client) and the client itself (after it's wired up)
 * @param {function} options.preProcessor A function that will be called with information
 *  about the client that WILL BE configured - name, memberName, url and config
 */
export default async function
configureServices(services, endpoints = {}, options = {}, logger = console) {
  const swaggerResourceCache = options.swaggerResources;
  const returnedServices = {};
  const workToDo = [];
  for (const [name, swaggerSpec] of Object.entries(services)) {
    const configOverride = endpoints[name];

    if (configOverride && 'enabled' in configOverride && !configOverride.enabled) {
      // So damn pedantic. I decree this use to be worthy.
      // eslint-disable-next-line no-continue
      continue;
    }

    // The value of the services key should be either a URL or a swagger specification
    // (not as a string, but as the hydrated javascript object)
    let specJson = swaggerSpec;
    let url;
    if (typeof swaggerSpec === 'string') {
      if (URL.parse(swaggerSpec).protocol) {
        // Remote reference
        url = swaggerSpec;
        specJson = undefined;
      } else {
        // Local file
        // eslint-disable-next-line no-await-in-loop
        specJson = await jsonResolver(swaggerSpec, null, swaggerResourceCache);
      }
    } else {
      // Resolve multi-document swagger clients against a configurable base path
      let refBase = options.basedir || path.resolve('.');
      if (configOverride && configOverride.baseReferencePath) {
        refBase = path.resolve(refBase, configOverride.baseReferencePath);
      }
      if (!refBase.endsWith(path.sep)) {
        refBase = `${refBase}${path.sep}`;
      }

      // eslint-disable-next-line no-await-in-loop
      specJson = await jsonResolver(specJson, refBase, swaggerResourceCache);
    }

    if (specJson) {
      if (typeof (configOverride) === 'string') {
        // Endpoint config is the URL of the service
        url = configOverride;
      } else if (configOverride) {
        // Endpoint config is input to URL.format with some tweaks
        const secScheme = configOverride.protocol || scheme(endpoints, swaggerSpec);
        const defaults = {
          hostname: name,
          protocol: secScheme,
        };
        defaults.port = secScheme.startsWith('https') ? 8443 : 8000;
        url = URL.format(Object.assign({}, defaults, configOverride));
        if (configOverride.security) {
          specJson.security = configOverride.security;
        }
      } else {
        // Use etcd/k8s for endpoint resolution by using the name of the service as a hostname
        const secScheme = scheme(endpoints, swaggerSpec);
        const defPort = secScheme.startsWith('https') ? 8443 : 8000;
        url = `${secScheme}//${name}:${defPort}`;
      }
    }

    // Sometimes it may be useful to override basePath and host directly
    if (specJson && configOverride && configOverride.basePath) {
      specJson.basePath = configOverride.basePath;
    }
    if (specJson && specJson.host && configOverride && configOverride.hostname) {
      specJson.host = configOverride.hostname;
    }
    if (specJson && specJson.schemes && configOverride && configOverride.protocol) {
      specJson.schemes = [configOverride.protocol.replace(':', '')];
    }

    if (specJson) {
      if (configOverride && configOverride.requestContentType) {
        specJson.consumes = [configOverride.requestContentType];
      } else if (!specJson.consumes || !specJson.consumes.length) {
        logger.warn('Missing "consumes" property on service, assuming application/json', { name });
        specJson.consumes = ['application/json'];
      }
    }

    logger.info('Connecting service', { name, url });
    const {
      username,
      authToken,
      securityScheme,
      password,
      swaggerOptions,
    } = configOverride || {};
    const clientConfig = Object.assign({
      url,
      spec: specJson,
      usePromise: true,
    }, options.swaggerOptions, swaggerOptions);
    if (username) {
      const schemeName = securityScheme || 'basic';
      clientConfig.authorizations = {
        [schemeName]: { username, password },
      };
    } else if (authToken) {
      clientConfig.authorizations = clientConfig.authorizations || {};
      clientConfig.authorizations[securityScheme || 'authToken'] = {
        token: { access_token: authToken, token_type: 'Bearer' },
      };
    }

    const workOrder = {
      url,
      name,
      memberName: _.upperFirst(_.camelCase(name)),
      config: clientConfig,
      endpoint: configOverride,
    };

    if (options.preProcessor) {
      // eslint-disable-next-line no-await-in-loop
      await options.preProcessor(workOrder, configOverride);
    }
    if (workOrder.config.spec && workOrder.config.url) {
      // Swagger will try to fetch this doc for refs if we set it
      delete workOrder.config.url;
    }
    workOrder.client = new Client(workOrder.config);
    workToDo.push(workOrder);
  }

  const clients = await Promise.all(workToDo.map(w => w.client));
  await Promise.all(clients.map(async (c, ix) => {
    const work = workToDo[ix];
    // Replace the promise with the real deal
    work.client = c;
    c.url = work.url;
    if (options.postProcessor) {
      await options.postProcessor(work);
    }
    work.client.configuredServiceName = work.memberName;
    for (const api of Object.values(work.client.apis)) {
      for (const method of Object.keys(api)) {
        const v2method = method.replace(/((_){2,})/g, '_')
          .replace(/^(_)*/g, '')
          .replace(/([_])*$/g, '');
        if (v2method !== method) {
          api[v2method] = api[method];
        }
      }
    }
    returnedServices[work.memberName] = work.client;
  }));
  return returnedServices;
}
